cesium/internal/domain/writer.go

```diff
@@ -139,31 +141,66 @@ func (w *Writer) Write(p []byte) (n int, err error) {
 // previous commit. If the provided timestamp is not strictly greater than the previous
 // commit, Commit will return an error. If the domain formed by the WriterConfig.Start
 // and the provided timestamp overlaps with any other domains within the DB, Commit will
-// return an error.
+// return an ErrDomainOverlap.
 func (w *Writer) Commit(ctx context.Context, end telem.TimeStamp) error {
 	ctx, span := w.T.Prod(ctx, "commit")
+	defer span.End()
+
 	if w.closed {
-		return span.EndWith(WriterClosedError)
-	}
-	if w.presetEnd {
-		end = w.End
+		return span.Error(WriterClosedError)
 	}
-	if err := w.validateCommitRange(end); err != nil {
-		return span.EndWith(err)
+
+	switchingFile, commitEnd, err := w.resolveCommitEnd(end)
+
+	if err = w.validateCommitRange(commitEnd, switchingFile); err != nil {
+		return span.Error(err)
 	}
 	length := w.internal.Len()
 	if length == 0 {
 		return nil
 	}
 	ptr := pointer{
-		TimeRange: telem.TimeRange{Start: w.Start, End: end},
+		TimeRange: telem.TimeRange{Start: w.Start, End: commitEnd},
 		offset:    uint32(w.internal.Offset()),
 		length:    uint32(length),
 		fileKey:   w.fileKey,
 	}
 	f := lo.Ternary(w.prevCommit.IsZero(), w.idx.insert, w.idx.update)
-	w.prevCommit = end
-	return span.EndWith(f(ctx, ptr))
+
+	if switchingFile {
+		err = w.internal.Close()
+		if err != nil {
+			return span.Error(err)
+		}
+
+		newFileKey, newInternalWriter, err := w.fc.acquireWriter(ctx)
+		if err != nil {
+			return span.Error(err)
+		}
+
+		w.fileKey = newFileKey
+		w.internal = newInternalWriter
+		w.Start = commitEnd
+		w.prevCommit = 0
+	} else {
+		w.prevCommit = commitEnd
+	}
+
+	return span.Error(f(ctx, ptr))
+}
+
+// resolveCommitEnd returns whether a file change is needed, the resolved commit end, and any errors.
+func (w *Writer) resolveCommitEnd(end telem.TimeStamp) (bool, telem.TimeStamp, error) {
+	s, err := w.fc.FS.Stat(fileKeyName(w.fileKey))
```

Comment:
I don't like calling `fs.Stat` here. We have an exclusive write lock on the file and we know how many bytes are written to the domain, so we know the size of the file. Why make the extra sys call? Also I don't like accessing the FS directly inside of the writer.

Reply:
Also why is this logic in `resolveCommitEnd`? It has nothing to do with resolving the commit end.

Reply:
Also this only checks the size of the file after we commit. This means that if we just kept writing data without committing, it would never roll over the file?